[WPE][GTK] WebProcess crashes at the end of every GLib API test

obyknovenius opened the pull request:
#### 4b27f17f5ade44bc02f5d8707c8140a762bb9253
<pre>
[WPE][GTK] WebProcess crashes at the end of every GLib API test
<a href="https://bugs.webkit.org/show_bug.cgi?id=259303">https://bugs.webkit.org/show_bug.cgi?id=259303</a>

Reviewed by Michael Catanzaro.

To run web process API tests we need to be able to execute code in
the web process. This is done by using web process extensions.

`WebProcessTest.cpp ` is a base class that implements the initialization
of a web process extension. This class also provides an API for
asserting that the `GObject` objects allocated in the web process during
test execution are deleted at the end of the test.

These `GObject` objects are stored in a static `HashSet`
`s_watchedObjects`. When the web process is terminating
`static void __attribute__((destructor)) checkLeaksAtExit()`
is called to verify that `s_watchedObjects` is empty, meaning
no objects were leaked.

The problem is that static variable destructors are invoking *before*
any `__attribute__((destructor))` functions. So by the time
`checkLeaksAtExit` is called, `s_watchedObjects` is already deallocated,
leading the web process to crash.

Because the web process extension modules we need for some tests are
located in the default directory, there are being loaded with every
API test. So we got the crash even if we don&apos;t run a web process test.

The solution to this problem is to wrap the watched objects hash set
into a static class and check for leaks in its destructor, before the
hash set is deallocated.

* Tools/TestWebKitAPI/Tests/WebKitGLib/DOMElementTest.cpp:
(DOMElementTest::testAutoFill):
* Tools/TestWebKitAPI/Tests/WebKitGLib/EditorTest.cpp:
(WebKitWebEditorTest::testSelectionChanged):
* Tools/TestWebKitAPI/Tests/WebKitGLib/FrameTest.cpp:
(WebKitFrameTest::testJavaScriptContext):
(WebKitFrameTest::testJavaScriptValues):
(WebKitFrameTest::testSubframe):
* Tools/TestWebKitAPI/Tests/WebKitGLib/WebProcessTest.cpp:
(runTest):
(webProcessTestRunnerFinalize):
(windowObjectClearedCallback):
(webkit_web_extension_initialize):
(checkLeaks): Deleted.
(WebProcessTest::assertObjectIsDeletedWhenTestFinishes): Deleted.
(checkLeaksAtExit): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitGLib/WebProcessTest.h:
(Watcher::~Watcher):
(Watcher::assertObjectIsDeletedWhenTestFinishes):
(Watcher::checkLeaks):
* Tools/TestWebKitAPI/Tests/WebKitGtk/AutocleanupsTest.cpp:
(AutocleanupsTest::testWebProcessAutocleanups):
* Tools/TestWebKitAPI/Tests/WebKitGtk/DOMClientRectTest.cpp:
(WebKitDOMClientRectTest::testDivBoundingClientRectPosition):
(WebKitDOMClientRectTest::testDivClientRectsPositionAndLength):
* Tools/TestWebKitAPI/Tests/WebKitGtk/DOMNodeFilterTest.cpp:
(WebKitDOMNodeFilterTest::testTreeWalker):
(WebKitDOMNodeFilterTest::testNodeIterator):
* Tools/TestWebKitAPI/Tests/WebKitGtk/DOMNodeTest.cpp:
(WebKitDOMNodeTest::testHierarchyNavigation):
(WebKitDOMNodeTest::testInsertion):
(WebKitDOMNodeTest::testTagNamesNodeList):
(WebKitDOMNodeTest::testTagNamesHTMLCollection):
(WebKitDOMNodeTest::testDOMCache):
* Tools/TestWebKitAPI/Tests/WebKitGtk/DOMXPathNSResolverTest.cpp:
(WebKitDOMXPathNSResolverTest::evaluateFooChildTextAndCheckResult):
(WebKitDOMXPathNSResolverTest::testXPathNSResolverNative):
(WebKitDOMXPathNSResolverTest::testXPathNSResolverCustom):

Canonical link: <a href="https://commits.webkit.org/266127@main">https://commits.webkit.org/266127@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d9b67879ba10115abceee21af9127cf9f5526970

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12977 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13297 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13628 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14715 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12381 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13045 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15802 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13322 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15077 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13143 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13842 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10977 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15166 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11130 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11732 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18794 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12205 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11898 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15108 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12371 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10260 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11634 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3177 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15952 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12212 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/WebProcessTest.cpp:
```diff
@@ -27,7 +27,40 @@
+        m_watchedObjects.add(object);
+        g_object_weak_ref(object, [](gpointer data, GObject* finalizedObject) {
+            Watcher* watcher = static_cast<Watcher*>(data);
+            watcher->m_watchedObjects.remove(finalizedObject);
```
Optional: I think the standard name for this parameter is slightly nicer:

```suggestion
        g_object_weak_ref(object, [](gpointer data, GObject* whereTheObjectWas) {
            Watcher* watcher = static_cast<Watcher*>(data);
            watcher->m_watchedObjects.remove(whereTheObjectWas);
```

obyknovenius replied:
Done.

mcatanzaro reviewed (approved):
Good solution.

Technically static objects are also destroyed in exit handlers, but I suppose they are destroyed sooner than the exit handler for the library destructor. Destructors of static objects and exit handlers are all pretty dangerous and we see crashes from them regularly, but in this case the benefit of being able to check for leaks is clearly worth the risk.

obyknovenius committed:
[WPE][GTK] WebProcess crashes at the end of every GLib API test
https://bugs.webkit.org/show_bug.cgi?id=259303

Reviewed by Michael Catanzaro.

To run web process API tests we need to be able to execute code in
the web process. This is done by using web process extensions.

`WebProcessTest.cpp ` is a base class that implements the initialization
of a web process extension. This class also provides an API for
asserting that the `GObject` objects allocated in the web process during
test execution are deleted at the end of the test.

These `GObject` objects are stored in a static `HashSet`
`s_watchedObjects`. When the web process is terminating
`static void __attribute__((destructor)) checkLeaksAtExit()`
is called to verify that `s_watchedObjects` is empty, meaning
no objects were leaked.

The problem is that static variable destructors are invoking *before*
any `__attribute__((destructor))` functions. So by the time
`checkLeaksAtExit` is called, `s_watchedObjects` is already deallocated,
leading the web process to crash.

Because the web process extension modules we need for some tests are
located in the default directory, there are being loaded with every
API test. So we got the crash even if we don't run a web process test.

The solution to this problem is to wrap the watched objects hash set
into a static class and check for leaks in its destructor, before the
hash set is deallocated.

* Tools/TestWebKitAPI/Tests/WebKitGLib/DOMElementTest.cpp:
(DOMElementTest::testAutoFill):
* Tools/TestWebKitAPI/Tests/WebKitGLib/EditorTest.cpp:
(WebKitWebEditorTest::testSelectionChanged):
* Tools/TestWebKitAPI/Tests/WebKitGLib/FrameTest.cpp:
(WebKitFrameTest::testJavaScriptContext):
(WebKitFrameTest::testJavaScriptValues):
(WebKitFrameTest::testSubframe):
* Tools/TestWebKitAPI/Tests/WebKitGLib/WebProcessTest.cpp:
(runTest):
(webProcessTestRunnerFinalize):
(windowObjectClearedCallback):
(webkit_web_extension_initialize):
(checkLeaks): Deleted.
(WebProcessTest::assertObjectIsDeletedWhenTestFinishes): Deleted.
(checkLeaksAtExit): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitGLib/WebProcessTest.h:
(Watcher::~Watcher):
(Watcher::assertObjectIsDeletedWhenTestFinishes):
(Watcher::checkLeaks):
* Tools/TestWebKitAPI/Tests/WebKitGtk/AutocleanupsTest.cpp:
(AutocleanupsTest::testWebProcessAutocleanups):
* Tools/TestWebKitAPI/Tests/WebKitGtk/DOMClientRectTest.cpp:
(WebKitDOMClientRectTest::testDivBoundingClientRectPosition):
(WebKitDOMClientRectTest::testDivClientRectsPositionAndLength):
* Tools/TestWebKitAPI/Tests/WebKitGtk/DOMNodeFilterTest.cpp:
(WebKitDOMNodeFilterTest::testTreeWalker):
(WebKitDOMNodeFilterTest::testNodeIterator):
* Tools/TestWebKitAPI/Tests/WebKitGtk/DOMNodeTest.cpp:
(WebKitDOMNodeTest::testHierarchyNavigation):
(WebKitDOMNodeTest::testInsertion):
(WebKitDOMNodeTest::testTagNamesNodeList):
(WebKitDOMNodeTest::testTagNamesHTMLCollection):
(WebKitDOMNodeTest::testDOMCache):
* Tools/TestWebKitAPI/Tests/WebKitGtk/DOMXPathNSResolverTest.cpp:
(WebKitDOMXPathNSResolverTest::evaluateFooChildTextAndCheckResult):
(WebKitDOMXPathNSResolverTest::testXPathNSResolverNative):
(WebKitDOMXPathNSResolverTest::testXPathNSResolverCustom):

Canonical link: https://commits.webkit.org/266127@main
